netebpfext/net_ebpf_ext_sock_addr.c

```diff
@@ -198,6 +295,246 @@ _net_ebpf_sock_addr_update_store_entries()
+    ExReleaseSpinLockExclusive(&_net_ebpf_ext_sock_addr_lock, old_irql);
+
+Exit:
+    return status;
+}
+
+static void
+_net_ebpf_ext_sock_addr_delete_redirect_handle(uint64_t filter_id)
+{
+    KIRQL old_irql;
+
+    old_irql = ExAcquireSpinLockExclusive(&_net_ebpf_ext_sock_addr_lock);
+
+    LIST_ENTRY* list_entry = _net_ebpf_ext_redirect_handle_list.Flink;
+    while (list_entry != &_net_ebpf_ext_redirect_handle_list) {
+        net_ebpf_extension_redirect_handle_entry_t* entry =
+            CONTAINING_RECORD(list_entry, net_ebpf_extension_redirect_handle_entry_t, list_entry);
+        if (entry->filter_id == filter_id) {
+            RemoveEntryList(list_entry);
+            ExFreePool(entry);
+            break;
+        }
+        list_entry = list_entry->Flink;
+    }
+    ExReleaseSpinLockExclusive(&_net_ebpf_ext_sock_addr_lock, old_irql);
+}
+
+static NTSTATUS
+_net_ebpf_ext_sock_addr_get_redirect_handle(uint64_t filter_id, _Out_ HANDLE* redirect_handle)
+{
+    NTSTATUS status = STATUS_NOT_FOUND;
+    KIRQL old_irql;
+
+    old_irql = ExAcquireSpinLockShared(&_net_ebpf_ext_sock_addr_lock);
+
+    LIST_ENTRY* list_entry = _net_ebpf_ext_redirect_handle_list.Flink;
+    while (list_entry != &_net_ebpf_ext_redirect_handle_list) {
+        net_ebpf_extension_redirect_handle_entry_t* entry =
+            CONTAINING_RECORD(list_entry, net_ebpf_extension_redirect_handle_entry_t, list_entry);
+        if (entry->filter_id == filter_id) {
+            *redirect_handle = entry->redirect_handle;
+            status = STATUS_SUCCESS;
+            break;
+        }
+        list_entry = list_entry->Flink;
+    }
+
+    ExReleaseSpinLockShared(&_net_ebpf_ext_sock_addr_lock, old_irql);
+
+    return status;
+}
+
+static void
+_net_ebpf_ext_initialize_connection_context(
+    bool redirected,
+    bool loopback,
+    uint32_t verdict,
+    uint64_t transport_endpoint_handle,
+    uint32_t protocol,
+    _Out_ net_ebpf_extension_connection_context_t* connection_context)
+{
+    connection_context->key.protocol = (uint16_t)protocol;
+    connection_context->value.loopback = loopback;
+    connection_context->value.redirected = redirected;
+    connection_context->value.verdict = (uint8_t)verdict;
+    connection_context->key.transport_endpoint_handle = transport_endpoint_handle;
+    connection_context->timestamp = KeQueryInterruptTime() / 10000; // in ms.
+}
+
+static bool
+_net_ebpf_ext_is_loopback_address(_In_ const bpf_sock_addr_t* address)
+{
+    SOCKADDR_STORAGE socket_address = {0};
+    socket_address.ss_family = (ADDRESS_FAMILY)address->family;
+    if (socket_address.ss_family == AF_INET) {
+        SOCKADDR_IN* v4_address = (SOCKADDR_IN*)&socket_address;
+        v4_address->sin_addr.S_un.S_addr = address->user_ip4;
+    } else {
+        SOCKADDR_IN6* v6_address = (SOCKADDR_IN6*)&socket_address;
+        RtlCopyMemory(v6_address->sin6_addr.u.Byte, address->user_ip6, 16);
+    }
+
+    return INETADDR_ISLOOPBACK((SOCKADDR*)&socket_address);
```

Comment:
Seems like this would be more simply
```suggestion
    return INET_IS_ADDR_LOOPBACK((ADDRESS_FAMILY)address->family, address->user_ip4);
```

Reply:
Actually removed this function totally